metrics-recorder-prometheus/src/lib.rs

```diff
@@ -29,80 +32,121 @@ impl PrometheusRecorder {
         self.output.push_str("\n# TYPE ");
-        self.output.push_str(label.as_str());
+        self.output.push_str(name.as_str());
         self.output.push_str(" counter\n");
-        self.output.push_str(label.as_str());
+        self.output.push_str(full_name.as_str());
         self.output.push_str(" ");
         self.output.push_str(value.to_string().as_str());
         self.output.push_str("\n");
     }
 
-    fn record_gauge<K: Into<Key>>(&mut self, key: K, value: i64) {
-        let label = key.into().as_ref().replace('.', "_");
+    fn record_gauge(&mut self, key: Key, value: i64) {
+        let (name, labels) = key_to_parts(key);
+        let full_name = render_labeled_name(&name, &labels);
         self.output.push_str("\n# TYPE ");
-        self.output.push_str(label.as_str());
+        self.output.push_str(name.as_str());
         self.output.push_str(" gauge\n");
-        self.output.push_str(label.as_str());
+        self.output.push_str(full_name.as_str());
         self.output.push_str(" ");
         self.output.push_str(value.to_string().as_str());
         self.output.push_str("\n");
     }
 
-    fn record_histogram<K: Into<Key>>(&mut self, key: K, values: &[u64]) {
-        let mut sum = 0;
-        let mut h = Histogram::<u64>::new(3).expect("failed to create histogram");
+    fn record_histogram(&mut self, key: Key, values: &[u64]) {
+        let entry = self.histos.entry(key).or_insert_with(|| {
+            let h = Histogram::<u64>::new(3).expect("failed to create histogram");
+            (0, h)
+        });
+
+        let (sum, h) = entry;
         for value in values {
             h.record(*value).expect("failed to record histogram value");
-            sum += *value;
-        }
-
-        let label = key.into().as_ref().replace('.', "_");
-        self.output.push_str("\n# TYPE ");
-        self.output.push_str(label.as_str());
-        self.output.push_str(" summary\n");
-
-        for quantile in &self.quantiles {
-            let value = h.value_at_quantile(quantile.value());
-            self.output.push_str(label.as_str());
-            self.output.push_str("{quantile=\"");
-            self.output.push_str(quantile.value().to_string().as_str());
-            self.output.push_str("\"} ");
-            self.output.push_str(value.to_string().as_str());
-            self.output.push_str("\n");
+            *sum += *value;
         }
-        self.output.push_str(label.as_str());
-        self.output.push_str("_sum ");
-        self.output.push_str(sum.to_string().as_str());
-        self.output.push_str("\n");
-        self.output.push_str(label.as_str());
-        self.output.push_str("_count ");
-        self.output.push_str(values.len().to_string().as_str());
-        self.output.push_str("\n");
     }
 }
 
 impl Clone for PrometheusRecorder {
     fn clone(&self) -> Self {
         Self {
             output: get_prom_expo_header(),
+            histos: HashMap::new(),
```

Comment:
Seems like an unexpected property for a clone impl? I think the average user would assume that everything in the recorder would get duplicated. I think you would probably want to use a differently named method.

Reply:
So, the `Clone` implementation here is related to #11 and the design of how recorders/exporters work together.

Basically, exporters are long-lived, and use a recorder to convert each `Snapshot` they take into the intended output format.  The idea was that you would create a "master copy" of a recorder with all of the configuration it needed, like display options, formatting options, whatever you want, and it would clone it every time it took a snapshot.

So in this case, the `Clone` implementation is specialized so that it keeps the configuration data -- the quantiles for displaying histograms -- but doesn't keep the actual rendered metric data.

From #11, my gut feeling is that a builder pattern is the right thing to do, I just haven't done it yet.

Thoughts?

Reply:
Yeah that really sounds like a mutable builder pattern. Because currently the fact that you don't actually use the "master copy" directly for recording, your instead use a clone "shallow copy" is quite confusing. I feel that by having a separate `RecorderBuilder` and `Recorder` trait you make that distinction clearer.

Other than that, what is the rationale of having short lived recorders? It seems like your would have performance benefits from having a recorder that can be reset after each snapshot, so that you don't have to allocate each time.

Reply:
For instance, your log recorder expects the recorder to implement `Into<String>` but I'm pretty sure your `Into<String>` implementation would have to allocate.

Reply:
The principal reason for preferring an owned recorder when processing a snapshot is that sharing a single recorder get very dicey with asynchronous callers, if not entirely impossible.  

Given the direction things are going with async/await and futures in Rust, I wanted `metrics` to be ahead of curve.  With that said, I'm not sure there's a place for a long-lived recorder that gets used over and over.

I'm definitely all for switching to a builder pattern, but I think I'd need to see a working example of a long-lived recorder being used from an asynchronous context (for example, using `metrics-exporter-http`) because I'm not sure it's actually possible.

Reply:
Gotcha. This would be definitely possible using a lower level polling library, but I'm not sure its achievable with current tiokio / futures 0.1. The `Stream` API is quite a mess imo.

I guess the real question is whether calling `Clone` is expensive enough that it would be worth it.